plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -504,32 +541,502 @@ synapticsmst_device_get_flash_checksum (SynapticsMSTDevice *device,
+						"Failed to read Tag");
+		}
+		else{
+			gboolean match = TRUE;
+			for (i=0; i<16; i++){
+				#if MSTDBG
+				g_debug("[%x]set tag vaild read %x tag %x", retryCount,readBuf[i], tagData[i]);
+				#endif
+				if (readBuf[i] != tagData[i]){
+					match = FALSE;
+				}
+			}
+			if (match){
+				break;
+			}
+		}
+	}while(retryCount++ < MAX_RETRY_COUNTS );
+	if( rc && retryCount== MAX_RETRY_COUNTS){
+				g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+									"set tag fail");
+	}
+	/* set tag invaild*/
+	if(rc == 0){
+		rc = synapticsmst_common_rc_get_command (connection,
+							 UPDC_READ_FROM_EEPROM,
+							 1, (EEPROM_BANK_OFFSET * BankInUse + EEPROM_TAG_OFFSET + 15), tagData);
+		if (rc) {
+			g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+					     "Failed to read from EEPROM of device");
+		}
+			
+		do
+		{
+			if (tagData[0] != 0xFF)	{  
+			/* CRC8 is not 0xFF, erase last 4k of bank# */
+				eraseOffset = (EEPROM_BANK_OFFSET * BankInUse + EEPROM_BANK_OFFSET - 0x1000) / 0x1000; // offset for last 4k of bank#
+				rc = synapticsmst_device_set_flash_sector_erase(device, FLASH_SECTOR_ERASE_4K, eraseOffset);				
+			}
+			else{  /* CRC8 is 0xFF, set it to 0x00 */
+				tagData[1] = 0x00;
+				rc = synapticsmst_common_rc_set_command (connection,
+								 UPDC_WRITE_TO_EEPROM,
+								 1, (EEPROM_BANK_OFFSET * BankInUse + EEPROM_TAG_OFFSET + 15), &tagData[1]);				
+			}
+
+			rc = synapticsmst_common_rc_get_command (connection,
+								UPDC_READ_FROM_EEPROM,
+								1, (EEPROM_BANK_OFFSET * BankInUse + EEPROM_TAG_OFFSET + 15), readBuf);
+			if (rc) {
+				g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+							"Failed to read from EEPROM of device");
+			}
+			else{
+				if ((readBuf[0] == 0xFF && tagData[0] != 0xFF) || (readBuf[0] == 0x00 && tagData[0] == 0xFF))				{
+					break;
+				}
+			}
+		}while( retryCount++ < MAX_RETRY_COUNTS );
+
+		#if MSTDBG
+		g_debug("Panamera update done");		
+		#endif
+		return TRUE;
+	}
+	else{
+		return FALSE;
+	}
+
+}
+
+
 gboolean
 synapticsmst_device_write_firmware (SynapticsMSTDevice *device,
 				    GBytes *fw,
 				    GFileProgressCallback progress_cb,
 				    gpointer progress_data,
 				    GError **error)
 {
+
 	const guint8 *payload_data;
-	guint32 payload_len;
+	guint32 payload_len, payload_len_max;
 	guint32 code_size = 0;
 	guint32 checksum = 0;
 	guint32 flash_checksum = 0;
 	guint32 offset = 0;
 	guint32 write_loops = 0;
 	guint32 data_to_write = 0;
+	guint32 dwData[16];
+	guint16 tmp;
+	guint16 erase_code = 0xFFFF;	
 	guint8 percentage = 0;
 	guint8 rc = 0;
-	guint16 tmp;
-	guint16 erase_code = 0xFFFF;
+	gboolean isChipPanamera = FALSE;
 	SynapticsMSTDevicePrivate *priv = GET_PRIVATE (device);
 	g_autoptr(SynapticsMSTConnection) connection = NULL;
 
+	if(synapticsmst_device_get_board_id (device)> 0x5000){
+		isChipPanamera = TRUE;
+		payload_len_max = PAYLOAD_SIZE_512K;
+		#if MSTDBG
```

Comment:
You can leave off the MSTDBG items here, `g_debug` will output this comment on it's own when `fwupd` or `fwupdtool` are launched in verbose.

Reply:
I already remove all of the MSTDBG.